It seemed to work the first time, but afterwards GTG crashes on startup with the following:

```
Traceback (most recent call last):
  File "/home/januz/code/gtg/.local_build/install/lib/python3.14/site-packages/GTG/gtk/application.py", line 121, in do_startup
    self.ds.find_and_load_file(data_file)
    ~~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^
  File "/home/januz/code/gtg/.local_build/install/lib/python3.14/site-packages/GTG/core/datastore.py", line 478, in find_and_load_file
    self.load_file(filepath)
    ~~~~~~~~~~~~~~^^^^^^^^^^
  File "/home/januz/code/gtg/.local_build/install/lib/python3.14/site-packages/GTG/core/datastore.py", line 212, in load_file
    self.load_data(self.xml_tree)
    ~~~~~~~~~~~~~~^^^^^^^^^^^^^^^
  File "/home/januz/code/gtg/.local_build/install/lib/python3.14/site-packages/GTG/core/datastore.py", line 194, in load_data
    self.tasks.from_xml(tasks_xml, self.tags)
    ~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^
  File "/home/januz/code/gtg/.local_build/install/lib/python3.14/site-packages/GTG/core/tasks.py", line 917, in from_xml
    assert added_element.text is not None, 'Added text not found in task '+str(tid)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AssertionError: Added text not found in task 945ff268-6976-5b56-9824-68ada97d91ec
Gtk-Message: 09:52:28.885: GtkDialog mapped without a transient parent. This is discouraged.
Traceback (most recent call last):
  File "/home/januz/code/gtg/.local_build/install/lib/python3.14/site-packages/GTG/gtk/application.py", line 161, in do_activate
    if self._check_exception():
       ~~~~~~~~~~~~~~~~~~~~~^^
  File "/home/januz/code/gtg/.local_build/install/lib/python3.14/site-packages/GTG/gtk/application.py", line 216, in _check_exception
    dialog.set_application(self)  # Keep application alive, for now
    ^^^^^^^^^^^^^^^^^^^^^^
AttributeError: 'NoneType' object has no attribute 'set_application'
Traceback (most recent call last):
  File "/home/januz/code/gtg/.local_build/install/lib/python3.14/site-packages/GTG/gtk/application.py", line 672, in do_shutdown
    self.ds.save()
    ~~~~~~~~~~~~^^
  File "/home/januz/code/gtg/.local_build/install/lib/python3.14/site-packages/GTG/core/datastore.py", line 250, in save
    assert path is not None, "Failed to determine save location."
           ^^^^^^^^^^^^^^^^
AssertionError: Failed to determine save location.

(gtg:679816): GLib-GIO-CRITICAL **: 09:52:28.916: GApplication subclass 'GTG+gtk+application+Application' failed to chain up on ::shutdown (from end of override function)

```

--- a/GTG/backends/backend_caldav.py
+++ b/GTG/backends/backend_caldav.py
@@ -23,6 +23,7 @@ Backend for storing/loading tasks in CalDAV Tasks
 import logging
 import re
 from collections import defaultdict
+from uuid import UUID
 from datetime import date, datetime
 from gettext import gettext as _
 from hashlib import md5
@@ -119,7 +120,9 @@ class Backend(PeriodicImportBackend):
             logger.warning("not loaded yet, ignoring set_task")
             return
         with self.datastore.mutex:
-            self.datastore.tasks.add(task)
+            if task.id not in self.datastore.tasks.lookup:
+                self.datastore.tasks.add(task)
+            # If task already exists, it was updated in-place by the backend
 
     @interruptible
     def remove_task(self, tid: str) -> None:
@@ -300,10 +303,9 @@ class Backend(PeriodicImportBackend):
             uid = UID_FIELD.get_dav(todo)
             self._cache.set_todo(todo, uid)
             # Updating and creating task according to todos
-            task = self.datastore.tasks.lookup[uid]
+            task = self.datastore.tasks.lookup.get(UUID(uid))
             if not task:  # not found, creating it
-                task = Task()
-                task.id = uid
+                task = Task(id=UUID(uid), title='')
                 Translator.fill_task(todo, task, self.namespace, self.datastore)
                 self.datastore.tasks.add(task)
                 counts['created'] += 1
@@ -337,7 +339,7 @@ class Backend(PeriodicImportBackend):
                 parents = PARENT_FIELD.get_dav(todo)
                 if (not parents  # no parent mean no relationship on build
                         or parents[0] in known_todos  # already known parent
-                        or self.datastore.tasks.lookup[uid]):  # already known uid
+                        or self.datastore.tasks.lookup.get(UUID(uid))):  # already known uid
                     yield todo
                     known_todos.add(uid)
             if loop >= MAX_CALENDAR_DEPTH:
@@ -348,8 +350,7 @@ class Backend(PeriodicImportBackend):
         """Getting all tasks that has the calendar tag"""
         for task in self.datastore.tasks.data:
             if CATEGORIES.has_calendar_tag(task, calendar):
-                # This line was broken in commit 74bd3f44 - see #1176
-                yield uid, task # noqa: F821
+                yield str(task.id), task
 
     #
     # Utility methods
@@ -407,9 +408,48 @@ class Field:
     def _is_value_allowed(self, value):
         return value not in self.ignored_values
 
+    # Mapping: old method names -> Task property names (new core)
+    PROPERTY_GET_MAP = {
+        'get_title':               'title',
+        'get_excerpt':             'content',
+        'get_uuid':                'id',
+        'get_start_date':          'date_start',
+        'get_due_date_constraint': 'date_due',
+        'get_closed_date':         'date_closed',
+        'get_added_date':          'date_added',
+        'get_modified':            'date_modified',
+    }
+
+    PROPERTY_SET_MAP = {
+        'set_uuid':        None,
+        'set_text':        'content',
+        'set_start_date':  'date_start',
+        'set_due_date':    'date_due',
+        'set_closed_date': 'date_closed',
+        'set_added_date':  'date_added',
+        'set_modified':    'date_modified',
+        'set_tags':        None,
+        'set_parent':      None,
+        'remove_parent':   None,
+    }
+
     def get_gtg(self, task: Task, namespace: str = None):
         "Extract value from GTG.core.task.Task according to specified getter"
-        return getattr(task, self.task_get_func_name)()
+        name = self.task_get_func_name
+        if name in self.PROPERTY_GET_MAP:
+            value = getattr(task, self.PROPERTY_GET_MAP[name])
+            if name == 'get_uuid':
+                return str(value)
+            return value
+        if name == 'get_tags_name':
+            return [t.name for t in task.tags]
+        if name == 'get_parents':
+            return [str(task.parent.id)] if task.parent else []
+        if name == 'get_children':
+            return [str(c.id) for c in task.children]
+        # Fallback: call as method (e.g. get_status)
+        attr = getattr(task, name)
+        return attr() if callable(attr) else attr
 
     def clean_dav(self, vtodo: iCalendar):
         """Will remove existing conflicting value from vTodo object"""
@@ -440,7 +480,18 @@ class Field:
 
     def write_gtg(self, task: Task, value, namespace: str = None):
         """Will write new value to GTG.core.task.Task"""
-        return getattr(task, self.task_set_func_name)(value)
+        name = self.task_set_func_name
+        if name in self.PROPERTY_SET_MAP:
+            prop = self.PROPERTY_SET_MAP[name]
+            if prop is not None:
+                setattr(task, prop, value)
+            return
+        # Fallback: call as method (e.g. set_title, set_status)
+        attr = getattr(task, name)
+        if callable(attr):
+            return attr(value)
+        else:
+            setattr(task, name, value)
 
     def set_gtg(self, todo: iCalendar, task: Task,
                 namespace: str = None) -> None:
@@ -648,7 +699,7 @@ class Categories(Field):
             task.add_tag(to_add)
         for to_delete in local_tags.difference(remote_tags):
             task.remove_tag(to_delete)
-        task.tags.sort(key=remote_tags.index)
+        # task.tags is a set in the new core, ordering is not supported
 
     def get_calendar_tag(self, calendar: iCalendar) -> str:
         return self.to_tag(calendar.name, DAV_TAG_PREFIX)

--- a/GTG/backends/backend_caldav.py
+++ b/GTG/backends/backend_caldav.py
@@ -43,6 +43,7 @@ logger = logging.getLogger(__name__)
 TAG_REGEX = re.compile(r'\B@\w+[-_()\w]*')
 MAX_CALENDAR_DEPTH = 500
 DAV_TAG_PREFIX = 'DAV_'
+REMOTE_UID_ATTR = 'remote_uid'
 
 # Set of fields whose change alone won't trigger a sync up
 DAV_IGNORE = {'last-modified',  # often updated alone by GTG
@@ -67,6 +68,21 @@ def uid_to_task_id(uid) -> UUID:
         return UUID(uid)
     except (TypeError, AttributeError, ValueError):
         return uuid5(UUID_NAMESPACE, str(uid))
+
+
+def remote_uid(task: Task, namespace: str) -> str:
+    """The UID the server knows this task by.
+
+    uid_to_task_id() is deliberately one-way (uuid5), so a task
+    imported from a server whose UIDs are not UUIDs cannot yield its
+    remote UID back from its id. Sending the GTG id in RELATED-TO
+    would point the server at a UID it never issued, and the
+    relationship would be silently dropped. The original UID is kept
+    as a task attribute at import time; tasks created in GTG are
+    pushed with their own id as UID, so falling back on it is correct.
+    """
+    return (task.get_attribute(REMOTE_UID_ATTR, namespace=namespace)
+            or str(task.id))
 
 
 class Backend(PeriodicImportBackend):
@@ -482,9 +498,11 @@ class Field:
         if name == 'get_tags_name':
             return [t.name for t in task.tags]
         if name == 'get_parents':
-            return [str(task.parent.id)] if task.parent else []
+            if not task.parent:
+                return []
+            return [remote_uid(task.parent, namespace)]
         if name == 'get_children':
-            return [str(c.id) for c in task.children]
+            return [remote_uid(child, namespace) for child in task.children]
         # Fallback: call as method (e.g. get_status)
         attr = getattr(task, name)
         return attr() if callable(attr) else attr
@@ -1105,6 +1123,7 @@ class Translator:
             task.add_tag(datastore.tags.new(to_add))
         for to_delete in local_tags.difference(remote_tags):
             task.remove_tag(to_delete)
+        task.set_attribute(REMOTE_UID_ATTR, UID_FIELD.get_dav(todo), **nmspc)
         task.set_attribute("url", str(todo.url), **nmspc)
         task.set_attribute("calendar_url", str(todo.parent.url), **nmspc)
         task.set_attribute("calendar_name", todo.parent.name, **nmspc)

--- a/GTG/backends/backend_caldav.py
+++ b/GTG/backends/backend_caldav.py
@@ -1097,6 +1097,14 @@ class Translator:
         if not CATEGORIES.has_calendar_tag(task, todo.parent):
             tag = datastore.tags.new(CATEGORIES.get_calendar_tag(todo.parent))
             task.add_tag(tag)
+        if not task.date_added:
+            # CREATED is optional in RFC 5545, and Task() starts with no
+            # added date. The core assumes every task has one: it
+            # serializes an empty <added> and then asserts on reload.
+            # Fall back on the modification date, which is always set,
+            # rather than 'now' which would claim the task was created
+            # during this import.
+            task.date_added = task.date_modified
         return task
 
     @classmethod

--- a/GTG/backends/generic_backend.py
+++ b/GTG/backends/generic_backend.py
@@ -654,12 +654,13 @@ class GenericBackend():
         """
         while not self.please_quit or bypass_quit_request:
             try:
-                task = self.to_set.pop()
+                tid = self.to_set.pop()
             except IndexError:
                 break
-            tid = task.id
             if tid not in self.to_remove:
-                self.set_task(task)
+                task = self.datastore.tasks.lookup.get(tid)
+                if task:
+                    self.set_task(task)
 
         while not self.please_quit or bypass_quit_request:
             try:
@@ -670,16 +671,15 @@ class GenericBackend():
         # we release the weak lock
         self.to_set_timer = None
 
-    def queue_set_task(self, task):
+    def queue_set_task(self, tid):
         """ Save the task in the backend. In particular, it just enqueues the
         task in the self.to_set queue. A thread will shortly run to apply the
         requested changes.
 
-        @param task: the task that should be saved
+        @param tid: the task UUID that should be saved
         """
-        tid = task.id
-        if task not in self.to_set and tid not in self.to_remove:
-            self.to_set.appendleft(task)
+        if tid not in self.to_set and tid not in self.to_remove:
+            self.to_set.appendleft(tid)
             self.__try_launch_setting_thread()
 
     def queue_remove_task(self, tid):

--- a/tests/backend/backend_caldav_test.py
+++ b/tests/backend/backend_caldav_test.py
@@ -469,6 +469,31 @@ class NonUuidUidRegressionTest(TestCase):
         self.assertEqual(0, counts2['deleted'])
         self.assertEqual(1, len(backend.datastore.tasks.lookup))
 
+    VTODO_NO_CREATED = ("BEGIN:VTODO\r\n"
+                        "DTSTAMP:20201212T172830Z\r\n"
+                        "LAST-MODIFIED:20201212T172558Z\r\n"
+                        "STATUS:NEEDS-ACTION\r\n"
+                        "SUMMARY:todo without a created date\r\n"
+                        "UID:no-created@example.org\r\n"
+                        "END:VTODO\r\n")
+
+    def test_import_todo_without_created_date(self):
+        """CREATED is optional in RFC 5545. A task imported without one
+        must still get an added date: the core serializes an empty
+        <added> and then refuses to reload the file."""
+        backend = self._backend()
+        calendar = Mock()
+        calendar.todos.return_value = [self._todo(self.VTODO_NO_CREATED)]
+        counts = {'created': 0, 'updated': 0, 'unchanged': 0, 'deleted': 0}
+        start = datetime.now(LOCAL_TIMEZONE)
+        backend._import_calendar_todos(calendar, start, counts)
+        self.assertEqual(1, counts['created'])
+        task = next(iter(backend.datastore.tasks.lookup.values()))
+        self.assertTrue(task.date_added,
+                        'task imported without CREATED has no added date, '
+                        'it will be saved as an empty <added> element')
+        self.assertNotEqual('', str(task.date_added))
+
     def test_extract_plain_text_with_subtask_reference(self):
         """Task content can reference subtasks as {!<task id>!} lines:
         extraction must resolve them through the new core API."""

--- a/tests/backend/backend_caldav_test.py
+++ b/tests/backend/backend_caldav_test.py
@@ -520,6 +520,15 @@ class NonUuidUidRegressionTest(TestCase):
         vtodo = Translator.fill_vtodo(child, 'My Calendar', backend.namespace)
         self.assertEqual('0', SORT_ORDER.get_dav(vtodo=vtodo.vtodo))
 
+    def test_export_related_to_carries_the_server_uid(self):
+        """RELATED-TO must carry the UID the server issued, not the GTG
+        id derived from it: uid_to_task_id() is one-way, so a mapped
+        uuid5 means nothing to the server and the link is dropped."""
+        backend, child = self._imported_child()
+        vtodo = Translator.fill_vtodo(child, 'My Calendar', backend.namespace)
+        self.assertEqual(['ROOT'], PARENT_FIELD.get_dav(vtodo=vtodo.vtodo),
+                         'the server would not recognize this parent UID')
+
     def test_calendar_tag_is_parsable_back_by_the_core(self):
         """A calendar named "Deck: Server" used to yield the tag
         DAV_Deck:_Server, which the editor re-read as @DAV_Deck and

